File: code/algo/getoffer/go/2/check_customdict_sort.go
package main

func isAlienSorted(words []string, order string) bool {
	index := [26]int{}
	for i, c := range order {
		index[c-'a'] = i
	}
next:
	for i := 1; i < len(words); i++ {
		for j := 0; j < len(words[i-1]) && j < len(words[i]); j++ {
			pre, cur := index[words[i-1][j]-'a'], index[words[i][j]-'a']
			if pre > cur {
				return false
			}
			if pre < cur {
				continue next
			}
		}
		if len(words[i-1]) > len(words[i]) {
			return false
		}
	}
	return true
}

File: code/algo/getoffer/c++/2/check_customdict_sort.cpp
#include <string>
#include <unordered_map>
#include <vector>

bool isAlienSorted(const std::vector<std::string> &words,
                   const std::string &order) {
  std::vector<int> index(26, 0);
  for (int i = 0; i < order.size(); ++i) {
    index[order[i] - 'a'] = i;
  }

  for (int i = 1; i < words.size(); ++i) {
    const std::string &prev = words[i - 1];
    const std::string &curr = words[i];
    for (int j = 0; j < std::min(prev.size(), curr.size()); ++j) {
      int pre = index[prev[j] - 'a'];
      int cur = index[curr[j] - 'a'];
      if (pre > cur) {
        return false;
      }
      if (pre < cur) {
        goto next_word;
      }
    }
    if (prev.size() > curr.size()) {
      return false;
    }
  next_word:;
  }
  return true;
}
